### ISSUE TITLE
Should we add a section describing how ISOBMFF fields need to be adjusted due to trimming?

### ISSUE THREAD

**Issue by shawnisingh** [AOMediaCodec/iamf#274]
If I understand, we have reached agreement that we would accept redundancy of trimming information in three places:  (1) within audio frame OBUs, (2) within the codec config describing Opus (when Opus is used), (3) using ELST and updating other timing information in ISOBMFF atoms.

I'm not aware of any spec that clearly documents all the moving parts that need to be updated in moov atoms, it's just implicitly described in ISOBMFF spec by the meaning of each field.

So, should we consider having an informative section that clarifies how the various ISOBMFF fields need to be adjusted to correctly represent trimming information?   The following fields may deserve being mentioned, though some of them the semantics would be that they are *unaffected* by the padding:

- ELST media time
- ELST segment duration
- mvhd duration
- mdhd duration
- sidx subsegment durations. (for fragmented mp4)
- tdft base media decode time (for fragmented mp4)
- trun last packet duration (and similar for non-fragmented)

Any other fields I may have missed?

**Comment by tdaede** [comment]
>     * ELST media time
>     * ELST segment duration

These are specified in 7.2.2 (edts) however we need to specify how to fill in the duration in edts.

>     * mvhd duration
>     * mdhd duration

We should do the same as the Opus spec (mdhd is pre-trimming, mvhd is post-trimming). Let's double check to be sure. @tdaede will check what demuxers use mdhd.

>     * sidx subsegment durations. (for fragmented mp4)
>     * tdft base media decode time (for fragmented mp4)

I believe these are pre-edit list and therefore don't need to be updated by the trimming section. We could explicitly specify this in the spec.

>     * trun last packet duration (and similar for non-fragmented)

Let's review the ISOBMFF spec and determine if this should be before or after trimming. This should be presentation time. (We can also check existing implementations).

We should also check if ctts is made mandatory by CMAF or other mapping specifications.

**Comment by sunghee-hwang** [comment]
[CTTS issue]
ISOBMFF spec says
```
The composition time to sample table is optional and shall only be present if DT and CT differ for any 
samples
```
In this point, ctts needs be present to indicate `sample_offset` for the samples having data to be trimmed at start.

Let me share my summary based on the discussion when we need to consider the playback of two files without gap.
One special case which we may need to consider is when to playback two adjacent files without gap. The last frame of the first file may have trimming data and/or the first frames of the second file may have trimming data.
For the gap-less playback of the two files, we may need to set as follows:
- stts: `sample_delta` for the last frame of the first file should be set to the duration after trimming. For example, when the last frame consists of 15ms audio samples followed by 5ms trimming data, `sample_delta` should set to 15ms.
- ctts: `sample_offset` for the first frames of the second file should be set to the duration for trimming. For example, (I assume that the first frame has only trimming data, in this example)  when the first frame consists of 10ms trimming data followed by 10ms audio samples, `sample_offset` should set to 10ms. 

Based on the above examples, 
- To playback of the first file: DT(1) = 0 and CT(1) = 10ms.
- For the gap-less playback of two files: When DT1(n=last frame) = 200ms and CT1(n) = 200ms, DT2(1) = 205 and CT2(1) = 215 (i.e. DT2(1) = DT1(n) + STTS1(n) - CTTS2(1) = 200ms + 15ms - 10ms = 205ms and CT2(1) = DT2(1) + STTS2(1) = 205 + 10 = 215) 

When we do not consider the gap-less playback, then the initial decoding time of the second file (DT2(0) would be 220ms but it should be 15ms (5ms stts + 10ms ctts) earlier than 220ms (i.e. 205ms) for the gap-less playback.

[Issue #108]
A following question is how to resolve issue the #108 (an entire audio frame is trimmed).
Actually, the data to be trimmed will not be shown in the presentation timeline. So, the audio frame can not have any composition time if the entire frame is trimmed).
```
- Samples (mdat): S1 (20ms trimming data) | S2 (10ms trimming data + 10ms audio samples) | S3 (20ms audio samples | ... | Sn (15ms audio samples + 5ms trimming data)
- Decoding Time (stts): S1 (DT(1) = 0ms) | S2 (DT(2) = 20ms) | S3 (DT(3) = 40ms) | ... | Sn (DT(n) = 200ms) 
  Where, DT(i) = DT(i-1) + STTS(i)
- Composition Time: S1 (CT(1)  = ??) | S2 (CT(2) = 30ms) | S3 (CT(3) = 40ms) | ... | Sn (CT(n) = 200ms)
  Where, CT(i) = DT(i) + CTTS(i) 
```
As AAC may have trimming data bigger than 1024, it may be difficult to restrict that IAMF does have no audio frame which the entire frame is trimmed.
So, we need to set a value to CT(1) anyway.
We may set to CT(1) = 20ms (i.e. `sample_offset` = 20ms) which implies that no data after trimming is presented at 20ms.


What do you think about this?

**Comment by sunghee-hwang** [comment]
> [CTTS issue] ISOBMFF spec says
> 
> ```
> The composition time to sample table is optional and shall only be present if DT and CT differ for any 
> samples
> ```
> 
> In this point, ctts needs be present to indicate `sample_offset` for the samples having data to be trimmed at start.
> 
> Let me share my summary based on the discussion when we need to consider the playback of two files without gap. One special case which we may need to consider is when to playback two adjacent files without gap. The last frame of the first file may have trimming data and/or the first frames of the second file may have trimming data. For the gap-less playback of the two files, we may need to set as follows:
> 
> * stts: `sample_delta` for the last frame of the first file should be set to the duration after trimming. For example, when the last frame consists of 15ms audio samples followed by 5ms trimming data, `sample_delta` should set to 15ms.
> * ctts: `sample_offset` for the first frames of the second file should be set to the duration for trimming. For example, (I assume that the first frame has only trimming data, in this example)  when the first frame consists of 10ms trimming data followed by 10ms audio samples, `sample_offset` should set to 10ms.
> 
> Based on the above examples,
> 
> * To playback of the first file: DT(1) = 0 and CT(1) = 10ms.
> * For the gap-less playback of two files: When DT1(n=last frame) = 200ms and CT1(n) = 200ms, DT2(1) = 205 and CT2(1) = 215 (i.e. DT2(1) = DT1(n) + STTS1(n) - CTTS2(1) = 200ms + 15ms - 10ms = 205ms and CT2(1) = DT2(1) + CTTS2(1) = 205 + 10 = 215)
> 
> When we do not consider the gap-less playback, then the initial decoding time of the second file (DT2(0) would be 220ms but it should be 15ms (5ms stts + 10ms ctts) earlier than 220ms (i.e. 205ms) for the gap-less playback.
> 
> [Issue #108] A following question is how to resolve issue the #108 (an entire audio frame is trimmed). Actually, the data to be trimmed will not be shown in the presentation timeline. So, the audio frame can not have any composition time if the entire frame is trimmed).
> 
> ```
> - Samples (mdat): S1 (20ms trimming data) | S2 (10ms trimming data + 10ms audio samples) | S3 (20ms audio samples | ... | Sn (15ms audio samples + 5ms trimming data)
> - Decoding Time (stts): S1 (DT(1) = 0ms) | S2 (DT(2) = 20ms) | S3 (DT(3) = 40ms) | ... | Sn (DT(n) = 200ms) 
>   Where, DT(i) = DT(i-1) + STTS(i)
> - Composition Time: S1 (CT(1)  = ??) | S2 (CT(2) = 30ms) | S3 (CT(3) = 40ms) | ... | Sn (CT(n) = 200ms)
>   Where, CT(i) = DT(i) + CTTS(i) 
> ```
> 
> As AAC may have trimming data bigger than 1024, it may be difficult to restrict that IAMF does have no audio frame which the entire frame is trimmed. So, we need to set a value to CT(1) anyway. We may set to CT(1) = 20ms (i.e. `sample_offset` = 20ms) which implies that no data after trimming is presented at 20ms.
> 
> NOTE: ISOBMFF define the definition of `sample' as follows:
> 
> ```
> 3.1.27
> sample
> all the data associated with a single time
> Note 1 to entry: No two samples within a track can share the same decoding time; no two samples can share the 
> same composition time.
> Note 2 to entry: In non-hint tracks, a sample is, for example, an individual frame of video, a series of video frames 
> in decoding order, or a compressed section of audio in decoding order; in hint tracks, a sample defines the 
> formation of one or more streaming packets.
> ```
> 
> What do you think about this?

I realize that I was wrong on the usage of `ctts` from "section 6.4 Time structure overview" of ISOBMFF spec. It says
```
Tracks consist of a sequence of samples in decoding order. Each sample has a decoding time that is 
computed by adding to the previous sample's decoding time, the previous sample's duration (as given 
by the values in the TimeToSampleBox or the equivalent field in movie fragments). The decoding time of 
the first sample is defined as being at time zero. This forms the decoding timeline of a track.
In some coding systems (notably video) samples are coded in a different order from their presentation 
order. In this case, each sample is assigned a composition time (as given by the values in the 
CompositionOffsetBox or the equivalent field in movie fragments), which is computed by adding a 
composition offset to the decoding time. This forms the composition timeline, and re-orders the samples 
into composition order.
```
As the audio frames are presented in decoding order, I feel that we don't need `ctts`. In other words, DT = CT.
Based on this, we can figure out whether `sample_delta` (stts, in moov) (or `sample_duration` (trun, in moof)) is the duration before or after trimming.
ISOBMFF spec says
- 1) The decoding time of the first sample is defined as being at time zero. (Section 6.4)
- 2) No two samples within a track can share the same decoding time. (Section 3.1.27)
- 3) The TimeToSampleBox contains decode time (DT) deltas: DT(n+1) = DT(n) + STTS(n) where STTS(n) is 
the (uncompressed) table entry for sample n. (Section 8.6.1.2.1)

If the first audio frame of a substream is entirely trimmed, which value should be set to STTS(1) (i.e. sample_delta for the first sample), 
- If the value should be after trimming, then STTS(1) = 0 so that DT(2) = DT(1). But it is a contradiction by 2)
So, `sample_delta` (or `sample_duration`) should be the duration before trimming.

In legacy implementation such as ffmpeg, the `sample_duration` is set to the duration of the last sample after trimming.
It is conflict with the above the definition. However, if we do not consider gapless playback of two files.
We can compromise the definition like this
- `sample_delta` (or `sample_duration`) is the duration of the sample before trimming if the sample is not the last sample.
- `edts` and `elst` boxes indicate the status of trimming.
- Gapless playback of two files is out of scope of IAMF.

In summary,
- IAMF does not need `ctts` (i.e. DT = CT)
- the definition of `sample_delta` is the duration before trimming except that of the last sample.
- `edts` and `elst` shall be mandatory for IAMF-OPUS and IAMF-AAC-LC.
- Gapless playback of two files is out of scope of IAMF.
NOTE: Players may support gapless playback from `edts` and `elst` boxes of each file.

How about this?

**Comment by yilun-zhangs** [comment]
- To playback of the first file: DT(1) = 0 and CT(1) = 10ms.
- For the gap-less playback of two files: When DT1(n=last frame) = 200ms and CT1(n) = 200ms, DT2(1) = 205 and CT2(1) = 215 (i.e. DT2(1) = DT1(n) + STTS1(n) - CTTS2(1) = 200ms + 15ms - 10ms = 205ms and CT2(1) = DT2(1) + CTTS2(1) = 205 + 10 = 215)

-->In my opinion, CT(1) = 10ms could not ensure the a/v sync, CT(1) should be -10ms. Of course, DT1(n) = 200ms, CT1(n) should be 190ms. (ISO spec, when version 1, CTTS could be signed)
i.e. DT2(1) = DT1(n) + STTS1(n) - CTTS2(1) 
--> I am confused, why should subtract CTTS2(1), on my understanding, the end decoding time of File1 is the start DTS of File2. So DT2(1) = DT1(n) + STTS1(n).

Any way, this has made the calculation be more complicated.
If we could reflect each packet's dts ( pts == dts) and actual duration after trimming, that is to say, output of IAMF decoder,
should be samples after trimming, it will be easier.

I checked the ISP spec, there is **sample_duration** in  'TRUN' box which could reflect duration.

For example, when the first frame  consists of 10ms audio samples and 10ms trimming data, the last frame consists of 15ms audio samples followed by 5ms trimming data.
so DT(1) = CT(1) = 0ms, duration is 10ms, DT(n) = CT(n) = 190ms, duration is 15ms.

The above is my opinion, if I am wrong, please understand it.
Thanks.

**Comment by sunghee-hwang** [comment]
If the sample_duration means after trimming, when the first frame is entirely trimmed, DT(1) = DT(2) which does not comply with the spec which says "No two samples within a track can share the same decoding time.". So, it should mean before trimming.

In our case, I think that DT = CT. And, then PT after trimming can be adjusted by the edit list from the CT.
Please refer to the below sentence of Section 6.4
```
The presentation timeline for each track is formed by the concatenation of sections of the composition 
timeline by means of explicit or implicit edit lists.
The presentation timelines of all the tracks are aligned at their zero point.
```
